File: Scala/data_structures/src/main/scala-2.12/heaps_and_sets/MergingTables.scala
package heaps_and_sets

object MergingTables {
  import scala.annotation.tailrec

  final case class TableOperation(destination: Int, source: Int)

  class UnionFind(nrTables: Int, initialTableRows: Seq[Int]) {
    private val tableRows: Array[Int] = initialTableRows.toArray
    private val parents: Array[Int] = (0 until nrTables).toArray
    private val ranks: Array[Int] = Array.fill(nrTables)(0)

    private def changeParentsToRoot(indicesOnPath: List[Int], root: Int): Unit =
      indicesOnPath.foreach(ix => parents(ix) = root)

    private def find(childIndex: Int): Int = {
      @tailrec
      def loop(id: Int, parentId: Int, indicesTowardsRoot: List[Int]): (Int, List[Int]) = {
        if (id == parentId) (id, indicesTowardsRoot)
        else loop(parentId, parents(parentId), id :: indicesTowardsRoot)
      }
      val (root, indicesOnPath): (Int, List[Int]) = loop(childIndex, parents(childIndex), Nil)
      changeParentsToRoot(indicesOnPath, root)
      root
    }

    def union(source: Int, destination: Int, largestTableSize: Int): Int = {
      val sourceId: Int = find(source)
      val destId: Int = find(destination)
      if (sourceId == destId) largestTableSize
      else if (ranks(sourceId) > ranks(destId)) {
        parents(destId) = sourceId
        tableRows(sourceId) += tableRows(destId)
        tableRows(destId) = 0
        math.max(largestTableSize, tableRows(sourceId))
      } else {
        parents(sourceId) = destId
        tableRows(destId) += tableRows(sourceId)
        tableRows(sourceId) = 0
        if (ranks(sourceId) == ranks(destId)) ranks(destId) += 1
        math.max(largestTableSize, tableRows(destId))
      }
    }
  }

  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  private def readTableOperations(reader: Iterator[String], nrOperations: Int): List[TableOperation] =
    (for { _ <- 0 until nrOperations } yield {
      val List(d, s): List[Int] = convertToIntList(reader.next())
      TableOperation(destination = d - 1, source = s - 1)
    }).toList

  def processMergeRequests(tableRows: List[Int], nrTables: Int, operations: List[TableOperation]): List[Int] = {
    val tables = new UnionFind(nrTables, tableRows)
    @tailrec
    def loop(ops: List[TableOperation], largestTableSize: Int, acc: List[Int]): List[Int] = ops match {
      case Nil => acc.reverse
      case TableOperation(destination, source) :: rest =>
        val updatedSize: Int = tables.union(source, destination, largestTableSize)
        loop(rest, updatedSize, updatedSize :: acc)
      }

    val initialMaxSize: Int = tableRows.max
    loop(operations, initialMaxSize, Nil)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val List(nrTables, nrOperations): List[Int] = convertToIntList(reader.next())
    val tableRows: List[Int] = convertToIntList(reader.next())
    val operations: List[TableOperation] = readTableOperations(reader, nrOperations)
    val result: List[Int] = processMergeRequests(tableRows, nrTables, operations)
    result.foreach(println)
  }
}

File: OCaml/data_structures/heaps_and_sets/merging_tables.ml
open Batteries

module type UnionFindSig =
    sig
        type t
        val initialize : int -> int list -> t
        val union : t -> int -> int -> int
    end

module UnionFind : UnionFindSig =
    struct
        type t = { table_rows : int array; parents : int array; ranks : int array }

        let largest_table_size = ref 0

        let change_parents_to_root parents indices_on_path root =
            List.iter (function ix -> parents.(ix) <- root) indices_on_path

        let initialize nr_tables initial_table_rows =
            largest_table_size := List.max initial_table_rows;
            { table_rows = Array.of_list initial_table_rows;
              parents = Array.init nr_tables (function ix -> ix);
              ranks = Array.make nr_tables 0 }

        let find parents child_ix =
            let rec loop id parent_id indices_towards_root =
                if id = parent_id then id, indices_towards_root
                else loop parent_id parents.(parent_id) (id :: indices_towards_root) in
            let root, indices_on_path = loop child_ix parents.(child_ix) [] in
            change_parents_to_root parents indices_on_path root;
            root

        let union { table_rows; parents; ranks } source destination =
            let source_id = find parents source in
            let dest_id = find parents destination in
            if source_id = dest_id then !largest_table_size
            else if ranks.(source_id) > ranks.(dest_id) then
                let () =
                    (parents.(dest_id) <- source_id;
                     table_rows.(source_id) <- table_rows.(source_id) + table_rows.(dest_id);
                     table_rows.(dest_id) <- 0;
                     largest_table_size := max !largest_table_size table_rows.(source_id))
                in !largest_table_size
            else
                let () =
                    (parents.(source_id) <- dest_id;
                     table_rows.(dest_id) <- table_rows.(dest_id) + table_rows.(source_id);
                     table_rows.(source_id) <- 0;
                     if ranks.(source_id) = ranks.(dest_id) then ranks.(dest_id) <- ranks.(dest_id) + 1;
                     largest_table_size := max !largest_table_size table_rows.(dest_id))
                in !largest_table_size
    end

type table_operation = { destination : int; source : int }


let convert_to_intlist line = List.map int_of_string Str.(line |> split (regexp " "))


let read_table_operations nr_operations =
    let rec read_one_line acc k =
        if k = nr_operations then List.rev acc
        else
            let d, s = Scanf.sscanf (read_line()) "%d %d" (fun x y -> x, y) in
            read_one_line ({ destination = d - 1; source = s - 1 } :: acc) (k + 1)
    in read_one_line [] 0


let process_merge_requests table_rows nr_tables operations =
    let tables = UnionFind.initialize nr_tables table_rows in
    List.map (function { destination; source } -> UnionFind.union tables source destination ) operations


let () =
    let nr_tables, nr_operations = Scanf.sscanf (read_line()) "%d %d" (fun x y -> x, y) in
    let table_rows = convert_to_intlist @@ read_line() in
    let operations = read_table_operations nr_operations in
    let results = process_merge_requests table_rows nr_tables operations in
    List.iter (function n -> print_int n; print_newline()) results
